### PR TITLE
feat: Allow writing holes which are solved by BRAT

### LINKED ISSUE #28: feat: Let BRAT fill in holes
This is one of the things implemented in the branch `new/inference-wip`. Add a new operator, `!` which is a hole in the program which BRAT will try to infer.

### PR DESCRIPTION
**Pull request by croyzor**
Closes #28 and #30

### CODE REVIEW

File: brat/Brat/Checker/SolvePatterns.hs [hunk 1]
```diff
@@ -186,45 +192,54 @@ unify l k r = do
...
+   -- Compute the value of the `tgt` variable from the known `src` value by inverting nv
+   tgtSrc <- invertNatVal nv
+   instantiateMeta (ExEnd src) (VNum (nVar (VPar (toEnd tgtSrc))))
+   wire (NamedPort src "", TNat, tgtSrc)
+
+ (ExEnd src, _) -> instantiateMeta (ExEnd src) (VNum nv)
+
+ -- Both targets, we need to create the thing that they both derive from
+ (InEnd bigTgt, [VPar (InEnd weeTgt)]) -> do
+   (_, [(idTgt, _)], [(idSrc, _)], _) <- anext "numval id" Id (S0, Some (Zy :* S0))
```
Comment (acl-cqc) [inline brat/Brat/Checker/SolvePatterns.hs:221]: (This case has no test coverage)

File: brat/Brat/Checker/SolvePatterns.hs [hunk 2]
```diff
@@ -266,9 +279,23 @@ unifyNum (NumValue lup lgro) (NumValue rup rgro)
   --   2^k * x
   -- = 2^k * (y + 1)
   -- = 2^k + 2^k * y
-  demandSucc (StrictMono k (Linear (VPar (ExEnd out)))) = do
-    y <- mkPred out
+  demandSucc _sm@(StrictMono _k (Linear (VPar (ExEnd _x)))) = error "Todo..." {-do
```
Comment (acl-cqc) [inline brat/Brat/Checker/SolvePatterns.hs:282]: I think the whole of `demandSucc` has no [test coverage](https://github.com/CQCL/brat/actions/runs/12122152161)

File: brat/Brat/Checker/SolvePatterns.hs [hunk 3]
```diff
@@ -287,53 +314,40 @@ unifyNum (NumValue lup lgro) (NumValue rup rgro)
...
+        solveNumMeta (ExEnd out) (n2PowTimes 1 (nVar (VPar (toEnd half))))
         pure (StrictMonoFun (StrictMono 0 (Linear (VPar (toEnd half)))))
-      Linear _ -> err . UnificationError $ "Can't force " ++ show n ++ " to be even"
+      Linear (VPar (InEnd tgt)) -> do
+        halfTgt <- buildNatVal (NumValue 0 (StrictMonoFun (StrictMono 1 (Linear (VPar (toEnd tgt))))))
+        let half = nVar (VPar (toEnd halfTgt))
+        solveNumMeta (InEnd tgt) (n2PowTimes 1 half)
+        pure (StrictMonoFun (StrictMono 0 (Linear (VPar (toEnd halfTgt)))))
       Full sm -> StrictMonoFun sm <$ demand0 (NumValue 0 (StrictMonoFun sm))
     evenGro (StrictMonoFun (StrictMono n mono)) = pure (StrictMonoFun (StrictMono (n - 1) mono))
```
Comment (acl-cqc) [inline brat/Brat/Checker/SolvePatterns.hs:326]: This is the only line of `evenGro` that [has test coverage](https://github.com/CQCL/brat/actions/runs/12122152161)

File: brat/Brat/Checker/SolvePatterns.hs [hunk 4]
```diff
@@ -287,53 +314,40 @@ unifyNum (NumValue lup lgro) (NumValue rup rgro)
...
 
     -- Check a numval is odd, and return its rounded down half
     oddGro :: Fun00 (VVar Z) -> Checking (NumVal (VVar Z))
     oddGro (StrictMonoFun (StrictMono 0 mono)) = case mono of
-      Linear (VPar (ExEnd out)) -> mkPred out >>= demandEven
-      Linear _ -> err . UnificationError $ "Can't force " ++ show n ++ " to be even"
+      -- TODO: Why aren't we using `out`??
```
Comment (acl-cqc) [inline brat/Brat/Checker/SolvePatterns.hs:331]: None of `oddGro` has test coverage

**Review by acl-cqc — COMMENTED** [review]
Ok, so
* The number of comments suggests you've managed to break out a good sized chunk of code that is not too overwhelming :-), which is an achievement in itself - well done and thank you :-)
* There's a lot of hard stuff in the nat-inversion-building that I really haven't followed through the logic - I can only say that it looks *plausible*. However really I would prefer to do one of
   * Add tests that cover all the cases and at least compile them down to Hugr. Then when we are able to execute them we suddenly gain a massive amount of assurance that we are actually building something sensible ;-) ;-)
   * If these functions are not covered by tests, then leave them `error "TODO"` here, and move the logic into another PR which can remain `WIP` until we do have tests (or at least, more time to think about them)
   * If the tests are only writable once we have 'Fork`, lets leave them `todo` and move the code into #41 (they are quite localized - no change to APIs - so there shouldn't be much fallout from leaving them out of this PR)

File: brat/Brat/Checker/Monad.hs [hunk 5]
```diff
@@ -50,12 +51,16 @@ data CtxEnv = CtxEnv
   , locals :: VEnv
   }
 
+type HopeSet = M.Map End FC
```
[the remark this answers is not in the document]
Comment (croyzor) [reply brat/Brat/Checker/Monad.hs:53]: The part that we care about (the keys of the map) are a set, the fact that they have FC information is an added bonus. We could have a `hopeSet :: Set End` and `hopeFCs :: Map End FC` where every End in `hopeSet` is a key in `hopeFCs`, but that seems a bit pointless.

Anyway, I see your point, I'll rename to `Hopes`

**Comment by croyzor** [comment]
Split into #69 and #70

**Pull request closed without merging**